### DISCUSSION TITLE
[HELP 🙏] SEVERE: Error instantiating a rule | java.lang.ClassNotFoundException

### DISCUSSION THREAD

**Discussion by felipeccruz** [pmd/pmd#4391] [Q&A] [answered]
I have been struggling to create a new rule in PMD for the past three days. I already followed all instructions from [#3463](https://github.com/pmd/pmd/discussions/3463), the templates from [pmd-examples](https://github.com/pmd/pmd-examples) and read the documentation.

When i execute pmd via command line with `pmd -d [path/to/file.java] -R [path/to/ruleset.xml]`, i get the following error:

![image](https://user-images.githubusercontent.com/73192682/218655101-6a03f1cc-42e4-48a2-a4c5-31e1f596a8c1.png)

I compiled the java rule linking the PMD lib as described:
`javac -cp "pmd-bin-6.54.0\lib\*" "src\main\java\CheckAbstractClassInheritanceRule.java"`

I generated the jar file and copied it into the pmd lib folder:
`jar --create --file CheckAbstractClassInheritanceRule.jar src\main\java\CheckAbstractClassInheritanceRule.class`

Seems like everything is correct, but i always get the same error.

My ruleset is like this, located at the same folder as the rules:

![image](https://user-images.githubusercontent.com/73192682/218654660-23e04176-c390-41f7-9cdc-e5dac550e806.png)

Could you help me, please? I really need this to run..

**Comment by adangel** [comment]
The problem might be as simple as wrong class name: In your ruleset, you are referencing `CheckAbstractClassInheritanceRule`. In your jar execution, you used a class called `WhileLoopsMustUseBracesRule`... So, you jar file contains the wrong class.

You can look into the jar file with `jar --list <name-of-the-file>.jar`.

**Reply by felipeccruz** [reply]
I'm sorry, i miswrote that line. The correct would be:

` jar --create --file CheckAbstractClassInheritanceRule.jar src\main\java\CheckAbstractClassInheritanceRule.class`

 So, the problem isn't at the jar execution.

Could you please check that again? I appreciate your attention.

**Reply by adangel** [reply]
OK, please have a look at the command on https://github.com/pmd/pmd-examples/tree/java-without-maven to create the jar file. Using your names, it should be

```
jar -c -f CheckAbstractClassInheritanceRule.jar -C src\main\java .
```

The parameter `-C` change the current directory before creating the jar file, so that files are added relatively.

The class file must be located at the root in the jar file. With you command, it was located at `src\main\java` - that would only be correct, if the java package would have been `src.main.java`... But you full class name (according to your ruleset) is `CheckAbstractClassInheritanceRule`, which is in the default package.

**Reply by felipeccruz** [reply]
I'm still not suceeding..
Here's exactly what i'm doing:

![image](https://user-images.githubusercontent.com/73192682/218882968-3e917b03-6706-4d5d-a723-e6bc0002c6d3.png)

I even copied the name, paths and commands from the example, but could not make it run.
What am i doing wrong?

**Reply by adangel** [reply]
Ok, so the problem is, that the batch file [pmd.bat](https://github.com/pmd/pmd/blob/master/pmd-dist/src/main/resources/scripts/pmd.bat) doesn't consider the environment variable `CLASSPATH`. It is only used by [run.sh](https://github.com/pmd/pmd/blob/master/pmd-dist/src/main/resources/scripts/run.sh) which is used for linux-like environments.

So, there are 3 options now:

1. The simplest is: Copy "custom-rule-example.jar" into the pmd's lib folder: `cp custom-rule-example.jar ..\..\pmd-bin-6.54.0\lib\`
2. Use a linux-like environment, such as [git bash](https://gitforwindows.org/)
3. Manually change the file `..\..\pmd-bin-6.54.0\bin\pmd.bat` to consider `CLASSPATH`, e.g.

```
@echo off
set TOPDIR="%~dp0.."
set OPTS=
set MAIN_CLASS=net.sourceforge.pmd.PMD
set PMD_CLASSPATH=%TOPDIR%\lib\*

if [%CLASSPATH%] NEQ [] (
    set PMD_CLASSPATH=%CLASSPATH%;%PMD_CLASSPATH%
)
java %PMD_JAVA_OPTS% -classpath %PMD_CLASSPATH% %OPTS% %MAIN_CLASS% %*
```

**Reply by jsotuyod** [reply]
For consistency it seems to me we should take the third approach as a definitive solution to this for future releases.

**Reply by adangel** [reply, marked as answer]
https://github.com/pmd/pmd/pull/4395

**Reply by felipeccruz** [reply]
Copying the "custom-rule-example.jar" into the pmd's lib folder doesn't seems to work. I tried it before and it gives me the same error.
I'll try the last one and give you feedback afterwards.

**Reply by adangel** [reply]
please try it again.
Also please post the output of the following command (if it doesn't work):

```
jar --list --file ..\..\pmd-bin-6.54.0\lib\custom-rule-example.jar
```

This should list the file "CheckAbstractClassInheritanceRule.class" at the root of the archive.

**Comment by jsotuyod** [comment]
How are you executing PMD? If the jar is located next to the ruleset, it may be a simple case of the jar not being included as part of the classpth when the jvm for PMD gets started.

Adding the jar to the lib folder of PMD will work for CLI. Under Maven you need to add a dependency under the plugin configuration. Gradle has a dedicated dependency set on which to add such dependencies.

**Reply by felipeccruz** [reply]
Hi! Yes, the .jar file is inside the pmd lib folder. The problem might be in the ruleset configuration, but i did it exactly like the [#3463](https://github.com/pmd/pmd/discussions/3463).

I'm executing pmd via command line with pmd -d [path/to/file.java] -R [path/to/ruleset.xml].
